File: src/main/javascript/bioassay.js
/*
 * This file contains methods to manipulate easily BioAssay Objects.
 *
 * @author Laurent Jourdren
 *
 * TODO add function to read/write imagene Files
 */

/*
 * Creating methods
 */

/*
 * Create an empty BioAssay object.
 * @param name The name of the bioAssay (optional)
 * @return a new bioAssay object
 */
function createBioAssay(name) {

  var nividicNames = JavaImporter();
  nividicNames.importPackage(Packages.fr.ens.transcriptome.nividic.om);

  with (nividicNames)  {
  
    var ba = BioAssayFactory.createBioAssay();
    
    if (name!=null) { ba.setName(name); }
    return ba;
  }

}

/*
 * Reader methods
 */

/*
 * Internal function.
 * Get the bioassay reader object
 * @param file File to read
 * @param type Type of the file to read
 * @return a BioAssayReader object
 */
function _getBioAssayReader(file, type) {

  var nividicNames = JavaImporter();
  nividicNames.importPackage(Packages.fr.ens.transcriptome.nividic.om.io);

  with (nividicNames)  {
    switch(type) {
  
      case "gpr": 
        return new GPRReader(file);

      case "idma": 
        return new IDMAReader(file);

      case "total.summary":
        return new TotalSummaryReader(file);
    }
  }

  return null;
}

/*
 * Read a bioAssay.
 * @param file File(s) to read
 * @param type Type of the file (gpr,idma...)
 * @param allFields Read all the fields or only default fields
 * @return a BioAssay file
 */
function readBioAssay(file, type, allFields, comma) {

  if (file==null) { return null; }
  
  if (file.constructor==String) { file = sf(file); }  

  if (file instanceof Array) {

    var result = new Array;
    for (i=0; i<file.length;i++) {
      
      var reader = _getBioAssayReader(file[i], type);

      if (allFields==true) { 
        reader.addAllFieldsToRead(); 
      }

      if (comma==true) {
        reader.setCommaDecimalSeparator(true);
      }

      result[i]=reader.read();
      result[i].setName(file[i].getName());
    }

    return result;
  } 
  else {

      var reader = _getBioAssayReader(file, type);

      if (allFields==true) { 
        reader.addAllFieldsToRead(); 
      }

      if (comma==true) {
        reader.setCommaDecimalSeparator(true);
      }

      var ba = reader.read();
      ba.setName(file.getName());
      return ba;
  }
}

/**
 * Shortcut to read GPR file(s).
 * @param file File(s) to read
 * @param allField read all fields
 * @return A BioAssay Object
 */
function readGPR(file, allFields) {

  return readBioAssay(file, "gpr", allFields, false);
}



/**
 * Shortcut to read IDMA file(s).
 * @param file File(s) to read
 * @param comma true if comma is the decimal separator
 * @return A BioAssay Object
 */
function readIDMA(file, comma) {

  return readBioAssay(file, "idma", false, comma);
}

/**
 * Shortcut to read total.summary file(s).
 * @param file File(s) to read
 * @param comma true if comma is the decimal separator
 * @return A BioAssay Object
 */
function readTotalSummary(file, comma) {

  return readBioAssay(file, "total.summary", false, comma);
}

/*
 * Writer methods
 */

/*
 * Internal function.
 * Get the bioassay reader object
 * @param file File to read
 * @param type Type of the file to read
 * @return a BioAssayReader object
 */
function _getBioAssayWriter(file, type) {

  if (file.constructor==String) { file = sf(file); }

  var nividicNames = JavaImporter();
  nividicNames.importPackage(Packages.fr.ens.transcriptome.nividic.om.io);

  with (nividicNames)  {
    switch(type) {
  
      case "gpr": 
        return new GPRWriter(file);

      case "idma": 
        return new IDMAWriter(file);

      case "total.summary":
        return new TotalSummaryWriter(file);
    }
  }

  return null;
}

/*
 * Write a bioAssay.
 * @param bioAssay BioAssay to write
 * @param file File(s) to write
 * @param type Type of the file (gpr,idma...)
 * @param allFields Write all the fields or only default fields
 * @return nothing
 */
function writeBioAssay(bioAssay, file, type, allFields) {

  var writer = _getBioAssayWriter(file, type);

  if (allFields==true) { 
    writer.addAllFieldsToWrite(); 
  }

  writer.write(bioAssay);
}

/**
 * Shortcut to write GPR file.
 * @param file File to write
 * @return nothing
 */
function writeGPR(bioAssay, file) {

  writeBioAssay(bioAssay, file, "gpr", true);
}

/**
 * Shortcut to write IDMA file.
 * @param file File to write
 * @return nothing
 */
function writeIDMA(bioAssay, file) {

  writeBioAssay(bioAssay, file, "idma", true);
}

/**
 * Shortcut to write total.summary file.
 * @param file File to write
 * @return nothing
 */
function writeTotalSummary(bioAssay, file) {

  writeBioAssay(bioAssay, file, "total.summary", true);
}

/*
 *  Sort methods
 */

/*
 * Get a sorter of bioAssay objects.
 * @return the sorter
 */
function createMASorter() {

  return new Packages.fr.ens.transcriptome.nividic.om.filters.BioAssayMASorterComparator();
}

/*
 * BioAssay FieldNames
 */
   
 /** Column name for red data. */
  var FIELD_NAME_RED = "red";
  /** Column name for green data. */
  var FIELD_NAME_GREEN = "green";
  /** Column name for flags data. */
  var FIELD_NAME_FLAG = "flags";
  /** Column name for name data. */
  var FIELD_NAME_ID = "id";
  /** Column name for ratio data. */
  var FIELD_NAME_RATIO = "ratio";
  /** Column name for bright data. */
  var FIELD_NAME_BRIGHT = "bright";
  /** Column name for description data. */
  var FIELD_NAME_DESCRIPTION = "description";
  /** Column name for a coordinate of a MA plot. */
  var FIELD_NAME_A = "a";
  /** Column name for m coordinate of a MA plot. */
  var FIELD_NAME_M = "m";
  /** Column name for the standard deviation of a values. */
  var FIELD_NAME_STD_DEV_A = "stddeva";
  /** Column name for the standard deviation of m values. */
  var FIELD_NAME_STD_DEV_M = "stddevm";

/*
 * Flags values
 */

  /** Flag bad. */
  var FLAG_BAD = -100;
  /** Flag abscent. */
  var FLAG_ABSCENT = -75;
  /** Flag not found. */
  var FLAG_NOT_FOUND = -50;
  /** Flag unflagged. */
  var FLAG_UNFLAGGED = 0;
  /** Flag normalized. */
  var FLAG_NORMALIZED = 1;
  /** Flag good. */
  var FLAG_GOOD = 100;
 

/*
 *  Filter methods
 */


/*
 * Create a filter on values greater or equals to the parameter.
 * @param field Field to use
 * @param threshold
 * @param condition
 */
function createThresholdFilter(field, threshold, condition) {

  if (field==null || threshold==null || condition ==null) return null;

  var nividicNames = JavaImporter();
  nividicNames.importPackage(Packages.fr.ens.transcriptome.nividic.om.filters);

  with(nividicNames) {

    /*obj = { test: function (v) { return !isNaN(v)  && v <= threshold; } };
    return new BioAssayMFilter(obj);*/
    
    return new BioAssayDoubleThresholdFilter(field, threshold, condition);
  }

}


/*
 * Create a filter on values greater or equals to the parameter.
 * @param threshold
 */
function createInfFilter(field,threshold) {

  return createThresholdFilter(field, threshold, "<=");
}

/*
 * Create a filter on values greater or equals to the parameter.
 * @param threshold
 */
function createSupFilter(field, threshold) {

  return createThresholdFilter(field, threshold, ">=");
}

/*
 * Create a filter on values greater or equals to the parameter.
 * @param threshold
 */
function createAInfFilter(threshold) {

  return createInfFilter(FIELD_NAME_A, threshold);
}

/*
 * Create a filter on values greater or equals to the parameter.
 * @param threshold
 */
function createASupFilter(threshold) {

  return createSupFilter(FIELD_NAME_A, threshold);
}


/*
 * Create a filter on values greater or equals to the parameter.
 * @param threshold
 */
function createMSupFilter(threshold) {

  return createSupFilter(FIELD_NAME_M, threshold);
}

/*
 * Create a filter on values greater or equals to the parameter.
 * @param threshold
 */
function createMInfFilter(threshold) {

  return createInfFilter(FIELD_NAME_M, threshold);
}

/*
 * Create a filter on values greater or equals to the parameter.
 * @param threshold
 */
function createAInfFilter(threshold) {

 return createInfFilter(FIELD_NAME_A, threshold);
}


/**
 * Swap the M values of a BioAssay.
 * @param bioAssay The bioAssay to swap
 * @return nothing
 */
function swapBioAssay(bioAssay) {

  var nividicNames = JavaImporter();
  nividicNames.importPackage(Packages.fr.ens.transcriptome.nividic.om);

  with(nividicNames) {
    
    BioAssayUtils.swap(bioAssay);
  }
}

File: src/main/java/fr/ens/transcriptome/nividic/om/BioAssay.java
/*
 *                      Nividic development code
 *
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public Licence.  This should
 * be distributed with the code.  If you do not have a copy,
 * see:
 *
 *      http://www.gnu.org/copyleft/lesser.html
 *
 * Copyright for this code is held jointly by the microarray platform
 * of the École Normale Supérieure and the individual authors.
 * These should be listed in @author doc comments.
 *
 * For more information on the Nividic project and its aims,
 * or to join the Nividic mailing list, visit the home page
 * at:
 *
 *      http://www.transcriptome.ens.fr/nividic
 *
 */

package fr.ens.transcriptome.nividic.om;

import fr.ens.transcriptome.nividic.om.filters.BioAssayFilter;
import fr.ens.transcriptome.nividic.om.filters.SpotComparator;

/**
 * This interface discribe a bioassay object. This code is based on previous
 * work in c++ developped at the Service de Génomique Fonctionelle of The
 * Commissariat à l'Énergie Atomique (CEA) by Laurent Jourdren, Miramme Allouche
 * and Vincent Frouin.
 * @author Laurent Jourdren
 * @author Vincent Frouin
 * @author Miriame Allouche
 */
public interface BioAssay extends BioAssayBase, BiologicalObject {

  /** Column name for red data. */
  String FIELD_NAME_RED = "red";
  /** Column name for green data. */
  String FIELD_NAME_GREEN = "green";
  /** Column name for flags data. */
  String FIELD_NAME_FLAG = "flags";
  /** Column name for name data. */
  String FIELD_NAME_ID = "id";
  /** Column name for ratio data. */
  String FIELD_NAME_RATIO = "ratio";
  /** Column name for bright data. */
  String FIELD_NAME_BRIGHT = "bright";
  /** Column name for description data. */
  String FIELD_NAME_DESCRIPTION = "description";
  /** Column name for a coordinate of a MA plot. */
  String FIELD_NAME_A = "a";
  /** Column name for m coordinate of a MA plot. */
  String FIELD_NAME_M = "m";
  /** Column name for the standard deviation of a values. */
  String FIELD_NAME_STD_DEV_A = "stddeva";
  /** Column name for the standard deviation of m values. */
  String FIELD_NAME_STD_DEV_M = "stddevm";

  /** Flag bad. */
  int FLAG_BAD = -100;
  /** Flag abscent. */
  int FLAG_ABSCENT = -75;
  /** Flag not found. */
  int FLAG_NOT_FOUND = -50;
  /** Flag unflagged. */
  int FLAG_UNFLAGGED = 0;
  /** Flag normalized. */
  int FLAG_NORMALIZED = 1;
  /** Flag good. */
  int FLAG_GOOD = 100;

  /** Idenfiers of the empty spots. */
  String[] EMPTY_SPOTS_IDENTIFIERS = {"", "empty"};

  //
  // Getters
  //

  /**
   * Get the annotations.
   * @return An annotations object
   */
  Annotation getAnnotation();

  /**
   * Get the reds data.
   * @return An int vector containing the data of the field
   */
  int[] getReds();

  /**
   * Get the green data.
   * @return An int vector containing the data of the field
   */
  int[] getGreens();

  /**
   * Get the identifiers data.
   * @return An int vector containing the data of the field
   */
  String[] getIds();

  /**
   * Test if the flags field exits.
   * @return <b>true </b> if the field exists.
   */
  int[] getFlags();

  /**
   * Get the array of description of the spots.
   * @return An array of string of the description.
   */
  String[] getDescriptions();

  /**
   * Get the array of the value of a coordinates of a MA plot.
   * @return An double array of a coordinate of a MA plot.
   */
  double[] getAs();

  /**
   * Get the array of the value of m coordinates of a MA plot.
   * @return An double array of m coordinate of a MA plot.
   */
  double[] getMs();

  /**
   * Get the array of the value of the standard deviations of A values.
   * @return An double array of the standard deviations of A values.
   */
  double[] getStdDevAs();

  /**
   * Get the array of the value of the standard deviations of M values.
   * @return An double array of the standard deviations of M values.
   */
  double[] getStdDevMs();

  /**
   * Test if the green field exits.
   * @return <b>true </b> if the field exists.
   */
  boolean isGreens();

  /**
   * Test if the red field exits.
   * @return <b>true </b> if the field exists.
   */
  boolean isReds();

  /**
   * Test if the ratio field exits.
   * @return <b>true </b> if the field exists.
   */
  boolean isRatios();

  /**
   * Test if the flags field exits.
   * @return <b>true </b> if the field exists.
   */
  boolean isFlags();

  /**
   * Get the brights data.
   * @return An int vector containing the data of the field
   */
  double[] getBrights();

  /**
   * Test if the bright field exits.
   * @return <b>true </b> if the field exists.
   */
  boolean isBrights();

  /**
   * Test if the identifier field exits.
   * @return <b>true </b> if the field exists.
   */
  boolean isIds();

  /**
   * Test if the description field is set.
   * @return true if the description field is set
   */
  boolean isDescriptions();

  /**
   * Test if the a field is set.
   * @return true if the a field is set
   */
  boolean isAs();

  /**
   * Test if the m field is set.
   * @return true if the m field is set
   */
  boolean isMs();

  /**
   * Test if the standard deviation for a values field is set.
   * @return true if the field is set
   */
  boolean isStdDevAs();

  /**
   * Test if the standard deviation for a values field is set.
   * @return true if the m field is set
   */
  boolean isStdDevMs();

  /**
   * Get the ratios data.
   * @return An int vector containing the data of the field
   */
  double[] getRatios();

  /**
   * Return some int vectors containing coordinates of the array spot.
   * @param metaRow Vector for meta lines
   * @param metaColumn Vector for meta columns
   * @param row Vector for lines
   * @param column Vector for columns
   * @throws BioAssayRuntimeException if there is no location or if one or more
   *           parameter is null.
   */
  void getLocations(int[] metaRow, int[] metaColumn, int[] row, int[] column)
      throws BioAssayRuntimeException;

  //
  // Setters
  //

  /**
   * Set the flags data.
   * @param flags The flags data to set
   * @throws BioAssayRuntimeException if flags is null or if flags size if wrong
   */
  void setFlags(int[] flags) throws BioAssayRuntimeException;

  /**
   * Set the green data.
   * @param greens The reds data to set *
   * @throws BioAssayRuntimeException if greens is null or if greens size if
   *           wrong
   */
  void setGreens(int[] greens) throws BioAssayRuntimeException;

  /**
   * Set the red data.
   * @param reds The reds data to set
   * @throws BioAssayRuntimeException if reds is null or if reds size if wrong
   */
  void setReds(int[] reds) throws BioAssayRuntimeException;

  /**
   * Set the ratios data.
   * @param ratios The ratios data to set
   * @throws BioAssayRuntimeException if ratios is null or if ratios size if
   *           wrong
   */
  void setRatios(double[] ratios) throws BioAssayRuntimeException;

  /**
   * Set the identifiers data.
   * @param ids The identifiers data to set
   * @throws BioAssayRuntimeException if ids is null or if ids size if wrong
   */
  void setIds(String[] ids) throws BioAssayRuntimeException;

  /**
   * Set the brights data.
   * @param brights The brights data to set
   * @throws BioAssayRuntimeException if brights is null or if brights size if
   *           wrong
   */
  void setBrights(double[] brights) throws BioAssayRuntimeException;

  /**
   * Set the description field.
   * @param descriptions Array of descriptions to set
   * @throws BioAssayRuntimeException if descriptions is null or if descriptions
   *           size if wrong
   */
  void setDescriptions(String[] descriptions) throws BioAssayRuntimeException;

  /**
   * Set the a field.
   * @param as Array of a coordinate to set
   * @throws BioAssayRuntimeException if as is null or if as size if wrong
   */
  void setAs(double[] as) throws BioAssayRuntimeException;

  /**
   * Set the m field.
   * @param ms Array of m coordinate to set
   * @throws BioAssayRuntimeException if ms is null or if ms size if wrong
   */
  void setMs(double[] ms) throws BioAssayRuntimeException;

  /**
   * Set the standard deviation a field.
   * @param stdDevAs Array of standard deviation for field m to set
   * @throws BioAssayRuntimeException if as is null or if as size if wrong
   */
  void setStdDevAs(double[] stdDevAs) throws BioAssayRuntimeException;

  /**
   * Set the standard deviation m field.
   * @param stdDevMs Array of standard deviation for field m to set
   * @throws BioAssayRuntimeException if ms is null or if ms size if wrong
   */
  void setStdDevMs(double[] stdDevMs) throws BioAssayRuntimeException;

  /**
   * Set the locations of the array with one vector for each coordinate.
   * @param metaRow Vector for meta lines
   * @param metaColumn Vector for meta columns
   * @param row Vector for lines
   * @param column Vector for columns
   * @return <b>true </b> if ok
   * @throws BioAssayRuntimeException if there is no location or if one or more
   *           parameter is null.
   */
  boolean setLocations(int[] metaRow, int[] metaColumn, int[] row, int[] column)
      throws BioAssayRuntimeException;

  //
  // Others methods
  //

  /**
   * Remove flags data.
   */
  void removeFlags();

  /**
   * Remove greens data.
   */
  void removeGreens();

  /**
   * remove reds data.
   */
  void removeReds();

  /**
   * remove ratios.
   */
  void removeRatios();

  /**
   * Remove identifiers.
   */
  void removeIds();

  /**
   * Remove brights.
   */
  void removeBrights();

  /**
   * Remove descriptions.
   */
  void removeDescriptions();

  /**
   * Remove a coordinate of MA plot.
   */
  void removeAs();

  /**
   * Remove m coordinate of MA plot.
   */
  void removeMs();

  /**
   * Get a spot iterator.
   * @return A spot Iterator Object.
   */
  SpotIterator iterator();

  /**
   * Get the spot at the index.
   * @param index The index of the spot
   * @return A spot object
   */
  Spot getSpot(int index);

  /**
   * Get the spot at the location.
   * @param location The location of the spot
   * @return A spot object
   */
  Spot getSpotLocation(int location);

  /**
   * Test if the spot is labeled as empty. If no SpotEmptyTester is set, use the
   * DefaultSpotEmptyTester.
   * @param index of the spot to test
   * @return true if the spot is labeled as empty
   */
  boolean isEmpty(int index);

  /**
   * Set the SpotEmptyTester.
   * @param spotEmptyTester SpotEmptyTester to set
   */
  void setSpotEmptyTester(SpotEmptyTester spotEmptyTester);

  /**
   * Get the SpotEmptyTester.
   * @return the SpotEmptyTester for the bioAssay
   */
  SpotEmptyTester getSpotEmptyTester();

  /**
   * Filter the bioAssay.
   * @param filter Filter to apply
   * @return a new bioAssay filtered
   */
  BioAssay filter(BioAssayFilter filter);

  /**
   * Count the entries of the bioAssay that pass the filter.
   * @param filter Filter to apply
   * @return a new bioAssay filtered
   */
  int count(BioAssayFilter filter);

  /**
   * Sort the bioAssay.
   * @param comparator Object used to do the sort
   * @return a new bioAssay sorted
   */
  BioAssay sort(SpotComparator comparator);

  /**
   * Swap Identifiers and Description columns
   */
  void swapIdentifiersAndDescriptions();

}